### PR TITLE
Leader not always dirty

### PR DESCRIPTION
**Pull request by goderbauer**
Prior to this change, `LeaderLayer`s had `alwaysNeedsAddToScene` unconditionally set to true, but they only need to always be added to scene _while_ actually connected to a FollowerLayer. This change makes it so and with that allows retained (= raster cachable) rendering of subtrees that contain `LeaderLayer`s, which are currently not connected to a follower. 

Change is motivated by customer "money": Fading in a route with a textfield was thrashing the raster cache because the textfield was adding a `LeaderLayer` that prevented retained rendering due to its `alwaysNeedsAddToScene` flag. The `LeaderLayer` is used to properly position toolbars and selection handles. However, when both are not visible (and their corresponding FollowerLayers are therefore not in the tree) there was no reason for this behavior. Raster caching the textfield (along any other content) during the fade transition is desirable for performance reason. This change also adds tests to ensure that a textfield without selection handles and toolbars visible remains raster-cachable.

Helps with b/203470598

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by LongCatIsLooong — APPROVED** [review]
LGTM but I'm not too familiar with how layers work. I think @Hixie originally wrote the leader/follower layers.

**Review by dnfield — COMMENTED** [review]
Layer changes LGTM

**Review by dnfield — APPROVED** [review]
Layer changes LGTM

**Comment by zanderso** [comment]
Internal benchmark results for this change: https://flutter-flutter-perf.skia.org/e/?begin=1635396934&end=1635464072&keys=X79fae4dec0e50b8f554a8a6f79f45812&num_commits=50&request_type=1&xbaroffset=26369

File: packages/flutter/lib/src/rendering/layer.dart [hunk 1]
```diff
@@ -2104,14 +2104,47 @@ class PhysicalModelLayer extends ContainerLayer {
...
+
+  /// Returns the [LeaderLayer] currently connected to this link.
+  ///
+  /// Valid in debug mode only. Returns null in all other modes.
+  ///
+  /// Use [registerFollower] to obtain the connected [LeaderLayer] for non-debug
+  /// purposes.
+  LeaderLayer? get debugLeader {
+    LeaderLayer? result;
+    if (kDebugMode) {
```
[the remark this answers is not in the document]
Comment (cchamm) [reply packages/flutter/lib/src/rendering/layer.dart:2132]: Some packages, like [local_hero](https://github.com/letsar/local_hero), depends on the `LayerLink.leader` field. 
Can it be exposed by a getter method again?
@goderbauer

[the remark this answers is not in the document]
Comment (goderbauer) [reply packages/flutter/lib/src/rendering/layer.dart:2132]: Re-exposed with the changes in https://github.com/flutter/flutter/pull/96486.

**Comment by creativecreatorormaybenot** [comment]
@goderbauer Making `_leader` prviate here is a breaking change - is this necessary for any reason?

We have a package that depends on this currently and I am wondering how I should handle it now. If there is a motivation for the change, could you explain that -  I would love to understand it.  
Otherwise, I would probably like to propose to open a PR to expose it again ~

**Comment by creativecreatorormaybenot** [comment]
> For Flutter a change is only breaking if it breaks one of the contributed tests (see also https://github.com/flutter/flutter/wiki/Tree-hygiene#1-determine-if-your-change-is-a-breaking-change). Since this change didn't require any changes to those tests it isn't considered breaking and no migration guide has been published. Please read the link above to learn how you can contribute tests to avoid being broken in the future.

@goderbauer This answer seems nice on the surface and I understand - but it is not always feasible to go through with the necessary overhead. Can you explain how to handle the breaking change?

**Comment by goderbauer** [comment]
The leader property has been exposed again as you can see in https://github.com/flutter/flutter/blob/28ca7bd7892bcb389208d0433022f45e88922841/packages/flutter/lib/src/rendering/layer.dart#L2113. 

To avoid that your package breaks on things like this in the future I highly recommend contributing your tests to https://github.com/flutter/tests.

**Comment by Hixie** [comment]
@creativecreatorormaybenot There shouldn't be much overhead if the package is already on GitHub... it's literally just a tiny file that specifies how to check out the package and run the tests. I'm very interested in hearing your experience if this is too much overhead, so that we can make it smoother.

**Comment by creativecreatorormaybenot** [comment]
@Hixie hey, I really appreciate it. I actually started to write a full repo of public-facing tests for the company I work(ed) at, but I ended up not publishing it because it was even too much overhead to finish it, basically needing to maintain another set of tests.

For packages, I fully agree. The `flutter_portal` package in particular does not have sufficient test coverage I would say, but if I have a package that qualifies, I will certainly contribute!

**Pull request merged**